### DISCUSSION TITLE
booting RO snapshots on EndeavourOS systems with Dracut instead of mkinitcpio?

### DISCUSSION THREAD

**Discussion by jfabernathy** [Antynea/grub-btrfs#277] [Q&A]
Since Endeavour OS switched to Dracut, I can't boot Read-only snapshots because I don't know how to configure the solution explained in one of the Read.md files. I had this working back when mkinitcpio was used.

I need some help as the explanation for Dracut systems doesn't make sense to me. Not sure where to modify the kernel parameters, or where to set the variable GRUB_BTRFS_SNAPSHOT_KERNEL_PARAMETERS

**Comment by Schievel1** [comment]
First off all I guess you need to tell dracut first, that you need to add the dracut modules for `overlayfs`. You can search in `/usr/lib/dracut/modules.d` for the overlayfs module, on gentoo it's called `90overlayfs`.
Add it to dracut using this: https://wiki.gentoo.org/wiki/Dracut#Adding_modules
Then you need to regenerate the initramfs of the kernel you want to boot the readonly snapshot with later. To regenerate every initramfs use `sudo dracut --regenerate-all -f -H `

I would suggest you first try passing this in Grub. So in Grub you go to the snapshot you want to boot, press e to edit the grub menu, navigate to where it says something like `linux /boot/vmlinuz-linux-6.2.11something root=UUID=somelongnonsese rw quiet splash` and add the `rd.live.overlay.overlayfs=1` (`rd.live.overlay.readonly=1`) to that line. 
Then press F10 to boot. 

However this is not permanent, it's only valid for this boot. And the nextime you boot you would have to type this in manually again. This is were the variable `GRUB_BTRFS_SNAPSHOT_KERNEL_PARAMETERS` in `/etc/grub.d/41_snapshots-btrfs` comes into play. This is simply a string that gets appended to every kernel command line of every snapshot by `grub-btrfs`. Set your read only argument there and regenerate your Grub menu. 

With all that said, I haven't got this done myself. I always wanted to test this out but I haven't found the time yet.

**Comment by jfabernathy** [comment]
Thanks for the information.  The version of grub-btrfs that is on the Archlinux repo is not up to date as far as I can tell.  I built grub-btrfs from source on one system and it's man grub-btrfs showed the GRUB_BTRFS_SNAPSHOT_KERNEL_PARAMETERS variable, but it didn't on my VM under test which was loaded from the repo.  

So I going to start over with a fresh VM and install grub-btrfs from the github version.  

I did confirm that the GRUB_BTRFS_SNAPSHOT_KERNEL_PARAMETERS variable get added to the /etc/default/grub-btrfs/config file before sudo /etc/grub.d/41_snapshot-btrfs script gets run.  The current readme implies this is all you have to do.

I'll test from scratch again.

**Reply by Schievel1** [reply]
Yes, that is all you have to do when it comes to grub-btrfs. You have to setup your system to run on an overlayfs of course. If I have gained a bit of experience with this I will try to improve the readme, but for now this is beyond the scope of grub-btrfs.

**Reply by jfabernathy** [reply]
my re-test indicates that the Cinnamon DE crashes when you try to boot a RO snapshot even with the overlay. So I'll try some other DEs.  It used to be that KDE Plasma didn't need the overlay to work correctly. Not sure now with Dracut

**Reply by Schievel1** [reply]
Hm but at least it seems to do something for you. I tried combinations of the different overlay kernel command line parameters for hour and rebooted my system over and over and every time I booted the snapshot was still writeable with persistent changes after reboot. 
can you tell me what parameters worked for you?

**Reply by jfabernathy** [reply]
My first re-test was KDE Plasma on EOS. Without any changes to any grub-btrfs file. it boot the snapper RO snapshots.  I could even boot to a pre snapshot and then run btrfs-assistant and restore that same snapshot I was booted on and even restored the matching @home snapshot. 

On the original test with Cinnamon DE, I use the GRUB_BTRFS_SNAPSHOT_KERNEL_PARAMETERS set to rd.live.overlay.overlayfs=1 and I also tried rd.live.overlay.readonly=1. They were on the kernel cmdline in grub, but Cinnamon crashed.

I've been told by Dalto on the EOS forum that KDE was safe for RO snapshot booting, but that was on mkinitcpio. It still works on Dracut

**Reply by jfabernathy** [reply]
I just did the test with EOS xfce4 and same results as Cinnamon.  RO snapshots will not boot with or without the GRUB_BTRFS_SNAPSHOT_KERNEL_PARAMETERS set. It tells me that something is missing. Probably the overlayfs module for dracut for Archlinux/EOS.

**Reply by jfabernathy** [reply]
According to the Arch wiki, "Overlayfs is enabled in the default kernel and the overlay module is automatically loaded upon issuing a mount command." Also, "rd.live.overlay.overlayfs=1" enables the use of the OverlayFS kernel module.

So according to the wiki nothing is missing and the kernel cmd option should work.

**Reply by Schievel1** [reply]
Yes the module is loaded but you have to have the module on your initramfs. When loaded into the initramfs your kernel has no access to the harddrive and it’s needs the module to get access to the harddrive. This is one of the things why initramfs were invented. 

Dracut does some auto detection and maybe included the module for you automatically. 

 But the snapshots get set back to their original state once you reboot, do they?

**Reply by jfabernathy** [reply]
The best I can tell Archlinux/eos does not have an overlayfs module in the available Dracut modules yet. There is no 90overlayfs or any version of "overlay" in /usr/lib/dracut/modules.d.  So I guess I'm stuck until that is added.

I tried the adding modules you referred me to from the gentoo wiki, but since the module does not exist in modules.d the sudo dracut --regenerate-all fails to find and install. I tried all the usual names list overlayfs, 90overlayfs, etc.

I guess I'm stuff for a while. Not critical at this point.

It is very strange to me how KDE-plasma on EOS can boot into snapshots without issue and does none of these things.

**Reply by Schievel1** [reply]
Yes they added this just recently, I think in dracut 58. See https://github.com/dracutdevs/dracut/pull/1934

But older dracut still supports overlayfs, ist just that this is in it's own module now. Before this was a part of the dmsquash-live module I think. 

I think when you do the regenerate-all thing, skip the -H argument that I suggested. This is kind of a gentooism, it lets dracut build a minimal initramfs that only includes the things the host needs. I think dracut fails to find a overlayfs module than since your current host is not using overlayfs. 

> It is very strange to me how KDE-plasma on EOS can boot into snapshots without issue and does none of these things.

So, again: you can boot into a overlayfs snapshot and it resets to the original state when rebooted into that snapshot? 
All that with an initramfs generated by dracut?

**Reply by jfabernathy** [reply]
Just checked the package for Archlinux for dracut and it's 056-2 and flagged out of date.

As to EOS KDE: Yes, I can pick a pre snapshot in the list on grub and boot it, login, and test to see the the application is missing as expected.  I can also use btrfs-assistant and restore the root and home snapshot to the very snapshot I'm actually booted to.  When I reboot to the normal kernel it's running the system without the app I had previously installed.

So it works just as I'd want with KDE and nothing special done to accommodate overlays parameters, etc.

I'll open a question on EOS forum about this strange behavior on KDE that works great.

**Reply by Schievel1** [reply]
Well but that means dracut works just fine for you. I don't thing the newer version with the overlayfs module will help with Cinnamon.

Regarding that, you have to have some directories in /var as their own subvolume for some desktop environments. Or the whole /var directory.

**Reply by jfabernathy** [reply]
I do have @cache on /var/cache, @log on /var/log and I think @tmp on /var/tmp

So somebody smart working on KDE must have put in the logic to deal with the RO snapshot without any of the grub-btrfs recommended modification.

I'd rather just stick with KDE but I'm on a 4K UHD display and the software I use for DVR playback has a few issues on Plasma.  I'm doing a lot of testing with Distros, DE, and snapper vs. timeshift.

Timeshift works very well with grub-btrfs, but there is no equivalent to snap-pac that properly does a pre-post snapshot that is properly labeled with what the snapshot was for. Just generic pre-post with no description. timeshift-autosnap isn't as good as snap-pac

**Reply by Schievel1** [reply]
At least for gnome you needed /var/lib/gdm and /var/lib/account service as well. I don't know how this is with cinnamon though.

**Comment by jfabernathy** [comment]
I have a solution at least for Endeavour OS (Archlinux). Copying some dracut modules from Garuda solve the issue. You don't need any of the grub-btrfs variable or kernel cmd line options. So you can use the pacman repo version (4.12-2) of grub-btrfs instead of the github master version here.

see [eos forum](https://forum.endeavouros.com/t/dracut-overlayfs-missing/39752/27)
for the details.

**Reply by Schievel1** [reply]
Thanks, that looks really promising. Any idea where this is coming from originally? Garuda maintainers?

From my understanding this script will activate overlayfs everytime when the initramfs is built with the module included, even when booting into the original system. Because it doesn't care for kernel parameters or anything at all. Is this correct?

**Reply by jfabernathy** [reply]
The person on the EOS forum helping me out said he stole the code off his Garuda system. If you checkout the link "eos forum" in my last post you'll  see more discussion about signing and cross signing issues with the dracut release that is supposed to have overlayfs included. Not sure what that will mean for how to setup booting read-only snapshots.  But it doesn't appear to need any kernel parameters in the snapshot grub entry like the grub-btrfs solution provides via /etc/grub.d/41_snapshot-btrfs and the variable uncommented in the grub-btrfs config file.

**Reply by Schievel1** [reply]
Thanks I will try this out as well.
Maybe we can include that in grub-btrfs in the future or something. It seems more written for the purpose than the original module of dracut, which is rather thought for live disks.

How does it determine if your want to boot into a overlayfs or boot for you though? When you boot your “real” system and not a snapshot, is it also overlayfs or normal persistent read-write?

**Reply by jfabernathy** [reply]
I'm too old to offer any read on code unless it's Fortran, but as I see the function mount_snapshot_overlay, it checks to see if the newly mounted root is type btrfs and it's read-only, then they mount a ram overlay.  Not sure about all that but clearly they are testing for a BTRFS RO file system being booted.

**Reply by Schievel1** [reply]
Yes, but your snapshots are not marked ro, are they?

**Reply by jfabernathy** [reply]
Default setup for snapper is RO snapshots.  When I look at btrfs-assistant table of snapshots I see all automatic ones as RO.

**Reply by jfabernathy** [reply]
![Screenshot from 2023-04-18 12-40-16](https://user-images.githubusercontent.com/46226222/232846225-44181cbb-785a-458f-85a0-8ecb841c6227.png)
